### PR TITLE
web: implement frame timings

### PR DESCRIPTION
**Pull request by yjbanov**
Implement `window.onReportTimings` for web.

### CODE REVIEW

File: lib/web_ui/test/canvaskit/frame_timings_test.dart [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import 'package:ui/ui.dart' as ui;
+
+import '../frame_timings_common.dart';
+
+void main() {
+  internalBootstrapBrowserTest(() => testMain);
+}
+
+void testMain() {
+  setUpAll(() async {
```
Comment (nturgut) [inline lib/web_ui/test/canvaskit/frame_timings_test.dart:18]: This test gives an error for ios-safari. [logs](https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket.appspot.com/8867598021470141728/+/steps/felt_ios-safari_test/0/stdout)

```
00:15 [32m+363[0m[33m ~157[0m[31m -1[0m: test/canvaskit/frame_timings_test.dart: (setUpAll) [1m[31m[E][0m[0m                                                                                                                             
  ReferenceError: Can't find variable: WebAssembly
  ../src/engine/canvaskit/initialization.dart 37:5 
```

Can we skip the test? Safari desktop tests are passing btw.

File: lib/web_ui/test/frame_timings_common.dart [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  ui.window.onDrawFrame = () {
+    final ui.SceneBuilder sceneBuilder = ui.SceneBuilder();
+    sceneBuilder
+      ..pushOffset(0, 0)
+      ..pop();
+    ui.window.render(sceneBuilder.build());
+    frameDone.complete();
+  };
+
+  // Frame 1
```
Comment (nturgut) [inline lib/web_ui/test/frame_timings_common.dart:29]: nit: I have told all comments have to end with a '.'  (same for frame 2)

File: lib/web_ui/test/frame_timings_common.dart [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  expect(timings, isNull, reason: '100 ms hasn\'t passed yet');
+  await Future<void>.delayed(const Duration(milliseconds: 150));
+
+  // Frame 2
+  frameDone = Completer<void>();
+  ui.window.scheduleFrame();
+  await frameDone.future;
+  expect(timings, hasLength(2), reason: '100 ms passed. 2 frames pumped.');
+  for (final ui.FrameTiming timing in timings) {
+    expect(timing.vsyncOverhead, greaterThanOrEqualTo(Duration.zero));
```
Comment (nturgut) [inline lib/web_ui/test/frame_timings_common.dart:41]: question: Is there a logical limit we can add? Or the value being more than zero is the only condition we are sure of.

File: lib/web_ui/lib/src/engine/profiler.dart [hunk 4]
```diff
@@ -107,3 +107,108 @@ class Profiler {
...
+
+/// Records the time when the framework started rasterizing the frame.
+///
+/// On the web, this value is almost always the same as [_buildFinishMicros]
+/// because it's single-threaded so there's no delay between building
+/// and rasterization.
+///
+/// This also means different things between HTML and CanvasKit renderers.
+///
+/// In HTML "rasterization" only captures DOM updates, but not work that
```
Comment (nturgut) [inline lib/web_ui/lib/src/engine/profiler.dart:168]: nit: 'the' work

File: lib/web_ui/lib/src/engine/profiler.dart [hunk 5]
```diff
@@ -107,3 +107,108 @@ class Profiler {
...
+  _rasterStartMicros = -1;
+  _rasterFinishMicros = -1;
+  if (now - _frameTimingsLastSubmitTime > _kFrameTimingsSubmitInterval) {
+    _frameTimingsLastSubmitTime = now;
+    window.invokeOnReportTimings(_frameTimings);
+    _frameTimings = <ui.FrameTiming>[];
+  }
+}
+
+/// Current timestamp in microseconds taken from the high-precision
```
Comment (nturgut) [inline lib/web_ui/lib/src/engine/profiler.dart:210]: I don't know how we are going to collect these values once we are running the tests. I remember we mentioned in the future we want to run them in Firefox as well. I found this comment for firefox in the [mozilla docs](https://developer.mozilla.org/en-US/docs/Web/API/Performance/now). We might want to keep it somewhere: "Firefox rounds the returned time to 1 millisecond increments."

Comment (yjbanov) [reply lib/web_ui/lib/src/engine/profiler.dart:210]: In the "Reduced time precision" section they explain how you can enable high precision timers. Worst case we just ignore data from Firefox if it's not useful.

Comment (yjbanov) [reply lib/web_ui/lib/src/engine/profiler.dart:210]: But good catch!

File: lib/web_ui/lib/src/engine/profiler.dart (hunk 4 above)
Comment (yjbanov) [reply lib/web_ui/lib/src/engine/profiler.dart:168]: Done.

File: lib/web_ui/test/frame_timings_common.dart (hunk 3 above)
Comment (yjbanov) [reply lib/web_ui/test/frame_timings_common.dart:41]: The test runs in real-time, so the actual number is unpredictable. Setting any upper limit may result in flakiness.

File: lib/web_ui/test/frame_timings_common.dart (hunk 2 above)
Comment (yjbanov) [reply lib/web_ui/test/frame_timings_common.dart:29]: Done.

File: lib/web_ui/test/canvaskit/frame_timings_test.dart (hunk 1 above)
Comment (yjbanov) [reply lib/web_ui/test/canvaskit/frame_timings_test.dart:18]: Oh, right! Done.

**Comment by yjbanov** [comment]
Responded to all comments. PTAL.

**Review by nturgut — APPROVED** [review]
LGTM. Thanks for adding me as a reviewer.

**Pull request merged**